File: easy_1/sieve/ruby/v2/sieve.rb
# sieve.rb

=begin
sieve = [2, 3, 4, 5, 6, 7, 8, 9, 10]
mock_sieve = [nil, nil, 2, 3, nil, 5, 6, 7, nil, 9, 10]
                           i
                                      m

primes_array = [2, 3]
limit = 10
idx = 3
mock_sieve[idx] = 2
multiplier = 3
=end


class Sieve
  def initialize(limit)
    @sieve = (2..limit).to_a
    @limit = limit
  end

  def primes
    primes_array = []
    mock_sieve = [nil, nil] + sieve

    2.upto(limit) do |int|
      if mock_sieve[int]
        multiplier = mock_sieve[int]
        primes_array << multiplier
        inner_idx = 2
        multiple = multiplier

        while multiple < limit
          multiple = multiplier * inner_idx
          mock_sieve[multiple] = nil
          inner_idx += 1
        end

        multiplier = nil
      end
    end

    primes_array
  end

  private
  attr_reader :limit, :sieve
end

File: easy_1/sieve/js/sieve.js
// sieve.js

class Sieve {
  constructor(limit) {
    this.limit = limit;
    this.sieve = this.createRangeFromTwoToLimitInclusive(limit);
  }

  createRangeFromTwoToLimitInclusive(limit) {
    const range = [];
    let n = 2;

    while (n <= limit) {
      range.push(n);
      n += 1;
    }

    return range;
  }

  primes() {
    const primesArray = [];
    const mockSieve = [null, null].concat(this.sieve);
    let idx = 2;

    while (idx < this.limit) {
      if (mockSieve[idx]) {
        primesArray.push(idx);
        let multiplier = idx;
        let multiple = idx;
        let innerIdx = 2;

        while (multiple < this.limit) {
          multiple = innerIdx * multiplier;
          mockSieve[multiple] = null;
          innerIdx += 1;
        }
      }

      idx += 1;
    }

    return primesArray;
  }
}

export default Sieve;